pyogrio/geopandas.py

```diff
@@ -226,13 +250,48 @@ def write_dataframe(
     # TODO: may need to fill in pd.NA, etc
     field_data = [df[f].values for f in fields]
 
-    if geometry_type is None:
-        geometry_type = "Unknown"
+    # Determine layer_geometry_type and/or promote_to_multi
+    if layer_geometry_type is None or promote_to_multi is None:
+        tmp_layer_geometry_type = "Unknown"
+
+        # If there is data, infer layer geometry type + promote_to_multi
         if not df.empty:
-            # TODO: validate geometry types, not all combinations are valid
             geometry_types = geometry.type.unique()
             if len(geometry_types) == 1:
-                geometry_type = geometry_types[0]
+                tmp_layer_geometry_type = geometry_types[0]
+                if promote_to_multi:
+                    if tmp_layer_geometry_type == "Polygon":
+                        tmp_layer_geometry_type = "MultiPolygon"
+                    elif tmp_layer_geometry_type == "LineString":
+                        tmp_layer_geometry_type = "MultiLineString"
+                    elif tmp_layer_geometry_type == "Point":
+                        tmp_layer_geometry_type = "MultiPoint"
+            elif len(geometry_types == 2):
+                # Check if the types are corresponding multi + single types
+                if "Polygon" in geometry_types and "MultiPolygon" in geometry_types:
+                    mixed_multi_single = "MultiPolygon"
+                elif (
+                    "LineString" in geometry_types
+                    and "MultiLineString" in geometry_types
+                ):
+                    mixed_multi_single = "MultiLineString"
+                elif "Point" in geometry_types and "MultiPoint" in geometry_types:
+                    mixed_multi_single = "MultiPoint"
+                else:
+                    mixed_multi_single = None
+
+                # If they are corresponding multi + single types
+                if (
+                    mixed_multi_single is not None
+                    and driver in DRIVERS_NO_MIXED_SINGLE_MULTI
+                ):
+                    if promote_to_multi is None:
```

Comment:
```suggestion
                    tmp_layer_geometry_type = multi_type if promote_to_multi is in (True None):
```

Reply:
promote_to_multi also needs to be set to True for the rest of the code to work properly